Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/VertexArrayMtl.mm

```diff
@@ -1092,7 +1092,11 @@ inline void SetDefaultVertexBufferLayout(mtl::VertexBufferLayoutDesc *layout)
     ANGLE_TRY(conversion->data.allocate(contextMtl, numVertices * targetStride, nullptr, &newBuffer,
                                         &newBufferOffset));
 
-    ANGLE_CHECK_GL_MATH(contextMtl, binding.getOffset() <= std::numeric_limits<uint32_t>::max());
+    GLintptr bindingOffset = binding.getOffset();
+
+    if constexpr (std::is_signed_v<GLintptr>)
+        ANGLE_CHECK_GL_MATH(contextMtl, bindingOffset >= 0);
```

Comment:
If `GLintptr` is signed on all platforms, then we can remove the `if constexpr` line.  (But if that's true, it's odd that the sign-compare warning doesn't occur on all platforms.)

Reply:
`GLintptr` should be signed, ptr-sized type.

Reply:
>it's odd that the sign-compare warning doesn't occur on all platforms
As far as I interpret the stackoverflow explanation:
- on 64-bit the compare is ` int64_t <= uint32_t` and compiler turns it to `int64_t <= int64_t`. This is sound comparison and doesn't warrant a warning.
- on 32-bit the compare is `int32_t <= uint32_t`, which compiler turns to `uint32_t <= uint32_t`. This dangerous comparison as it inverts the result for negative numbers, and as such it should be warning.

Reply:
@kkinnunen-apple Via Slack DM you said this change was "wrong", but I'm not sure I understand "how" you think it's wrong.

Is it that this change should only exist on 32-bit systems and should not apply to 64-bit systems?

My assumption was that even if the check wasn't _strictly_ necessary for 64-bit systems, it was still a valid check since a `signed` value could still be negative, and existing code is just happily casting a signed value to an unsigned value without checking whether it's greater than or equal to zero.

Reply:
> I'm not sure I understand "how" you think it's wrong.

Just checking, did you notice the explanations above? I did try to explain the rationale.

Is it that this change should only exist on 32-bit systems and should not apply to 64-bit systems?

I think most of them are incorrect "in spirit" on 32-bit systems and unneeded on 64-bit systems. 

The logic is: just because the intptr type is signed, it does not mean that it should be compared < 0 before converting to a 32-bit value for an algorithm that has 32 bit limitation.

So as an example: the binding offset here comes from the GL API, `pointer` argument:
`void VertexAttribPointer( uint index, int size, enum type, boolean normalized, sizei stride, const void *pointer );`

If the caller wants to specify 2GB of vertex attrib offset, they say `reinterpret_cast<void*>(0x80000000)`.

Now, suppose the previous code was:
```
uint32_t offset  = static_cast<uint32_t>( binding.getOffset());
```

So you cannot just blanket change that to:
```
ANGLE_CHECK_GL_MATH(contextMtl, binding.getOffset() >= 0);
uint32_t offset  = static_cast<uint32_t>( binding.getOffset());
```

If you do this, you might break some case that was theoretically working before.

This is because it's not a signed value to begin with -- it is just stored in a signed type for various convoluted reasons.

So to fix the signedness warnings you need to fix them in a way that preserves the semantics of the value being used.

So in this binding case:
```
ANGLE_CHECK_GL_MATH(contextMtl, binding.getOffset() <= std::numeric_limits<uint32_t>::max());
```

This is a valid check, as it appears that the algorithm is limited to 32 bits.

It's a pointless check on 32-bit platforms, as the GLintptr_t, int, is promoted to unsigned and <= max unsigned, which is a tautology.

It's triggering a signedness comparison warning, as it is promoting int to unsigned.

The warning should be fixed in a way that preserves the existing algorithm, unless of course the existing algorithm is broken.

Reply:
Thanks for the explanation!  I think I understand now.